gpg_key.go

```diff
@@ -0,0 +1,184 @@
+	// Uploads a GPG Key to a private registry scoped with a namespace.
+	Create(ctx context.Context, options GPGKeyCreateOptions) (*GPGKey, error)
+
+	// Read a GPG key.
+	Read(ctx context.Context, keyID GPGKeyID) (*GPGKey, error)
+
+	// Update a GPG key.
+	Update(ctx context.Context, keyID GPGKeyID, options GPGKeyUpdateOptions) (*GPGKey, error)
+
+	// Delete a GPG key.
+	Delete(ctx context.Context, keyID GPGKeyID) error
+}
+
+// gpgKeys implements GPGKeys
+type gpgKeys struct {
+	client *Client
+}
+
+// GPGKey represents a signed GPG key for a TFC/E private provider.
+type GPGKey struct {
+	ID             string    `jsonapi:"primary,gpg-keys"`
+	AsciiAmor      string    `jsonapi:"attr,ascii-armor"`
+	CreatedAt      time.Time `jsonapi:"attr,created-at,iso8601"`
+	KeyID          string    `jsonapi:"attr,key-id"`
+	Namespace      string    `jsonapi:"attr,namespace"`
+	Source         string    `jsonapi:"attr,source"`
+	SourceURL      *string   `jsonapi:"attr,source-url"`
+	TrustSignature string    `jsonapi:"attr,trust-signature"`
+	UpdatedAt      time.Time `jsonapi:"attr,updated-at,iso8601"`
+}
+
+// GPGKeyID represents the set of identifiers used to fetch a GPG key.
+type GPGKeyID struct {
+	Namespace string
+	KeyID     string
+}
+
+// GPGKeyCreateOptions represents all the available options used to create a GPG key.
+type GPGKeyCreateOptions struct {
+	Type      string `jsonapi:"primary,gpg-keys"`
+	Namespace string `jsonapi:"attr,namespace"`
+	AsciiAmor string `jsonapi:"attr,ascii-armor"`
+}
+
+// GPGKeyCreateOptions represents all the available options used to update a GPG key.
+type GPGKeyUpdateOptions struct {
+	Type      string `jsonapi:"primary,gpg-keys"`
+	Namespace string `jsonapi:"attr,namespace"`
+}
+
+func (s *gpgKeys) Create(ctx context.Context, options GPGKeyCreateOptions) (*GPGKey, error) {
+	if err := options.valid(); err != nil {
+		return nil, err
+	}
+
+	req, err := s.client.newRequest("POST", "/api/registry/private/v2/gpg-keys", &options)
+	if err != nil {
+		return nil, err
+	}
+
+	g := &GPGKey{}
+	err = s.client.do(ctx, req, g)
+	if err != nil {
+		return nil, err
+	}
+
+	return g, nil
+}
+
+func (s *gpgKeys) Read(ctx context.Context, keyID GPGKeyID) (*GPGKey, error) {
+	if err := keyID.valid(); err != nil {
+		return nil, err
+	}
+
+	u := fmt.Sprintf("/api/registry/private/v2/gpg-keys/%s/%s",
+		url.QueryEscape(keyID.Namespace),
+		url.QueryEscape(keyID.KeyID),
+	)
+	req, err := s.client.newRequest("GET", u, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	g := &GPGKey{}
+	err = s.client.do(ctx, req, g)
+	if err != nil {
+		return nil, err
+	}
+
+	return g, nil
+}
+
+func (s *gpgKeys) Update(ctx context.Context, keyID GPGKeyID, options GPGKeyUpdateOptions) (*GPGKey, error) {
+	if err := options.valid(); err != nil {
+		return nil, err
+	}
+
+	if err := keyID.valid(); err != nil {
+		return nil, err
+	}
+
+	u := fmt.Sprintf("/api/registry/private/v2/gpg-keys/%s/%s",
+		url.QueryEscape(keyID.Namespace),
```

Comment:
When I was reading the [docs](https://www.terraform.io/cloud-docs/api-docs/private-registry/gpg-keys#update-a-gpg-key), I noticed this line: `Only the namespace attribute can be updated, and namespace has to match an organization the user has permission to access.` I'm curious if the API returns an error if a namespace doesn't match an org the user has permission to access? Would this be something we would want to add a test for? The reason I suggest adding a test is because this detail could be overlooked by someone using this API, and they might see a test. I also admit that we don't want to cross a boundary if it doesn't make sense. What do you think? 🤔

Reply:
Good catch and this makes way more sense here as we want to test for an error that's specific to this resource! I've added a check in the error handler for `client.do()` inside `GPGKeys.Update()` where if the body's message contains `namespace is not authorized` it will return `ErrNamespaceNotAuthorized`. I've also added the corresponding test.